Modules API

nilsonragee opened the pull request:
Как мы знаем, все наши `Vulkan`-овские модули внутри `ref/vk` так или иначе имеют концепцию инициализации и деинициализации посредством вызова функций:
- `xxx_Init()` - для инициализации модуля перед тем, как его использовать.
- `xxx_Shutdown()` - для деинициализации модуля после того, как им закончили пользоваться.
 
Данная концепция подразумевает, что перед тем, как использовать что-то из нужного нам модуля, мы сначала должны этот модуль инициализировать, а когда он нам не нужен, то мы вызываем деинициализацию, чтобы этот модуль сделал все необходимые действия для плавного возвращения всех ресурсов, которые он занял, обратно в систему.

Реализовано это у нас очень просто, но эта простота упускает реализацию достаточно важных деталей:
1. ### Состояние инициализации
Мы его просто напросто не знаем. То есть каждый раз перед тем, как использовать функционал модуля, мы должны либо инициализировать его, даже если он уже инициализирован, так как мы не знаем об этом, либо надеяться на то, что кто-то уже сделал это до нас.

2. ### Зависимости 
Все они задаются вручную и никак не отслеживаются самим модулем. Нет какого-то определенного списка зависимостей, который мы просто можем взять и посмотреть, а также, чтобы это сделал сам модуль и проинициализировал их, если это еще не было сделано. "Списки" этих зависимостей нигде не хранятся, кроме как в самом коде в виде вызовов их функций инициализации, которые обычно идут списком внутри функции `xxx_Init()`.

Этот PR является попыткой все это реорганизовать и вынести в отдельный, доведенный до ума API. Намеки на надобность такого рода API уже есть комментариях самого кода: 

https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_core.c#L680-L707

Осталось лишь понять как именно этот API реализовать и что конкретно мы от него хотим. Я выделил следующие критерии (за исключением всяких "простота в использовании", "понятность" и т.д. - берем это, как базовые критерии):
1. Задание зависимостей - чтобы мы вручную не прописывали все `Init`-ы внутри других `Init`-ов.
2. Автоматическая иниациализация всех заданных зависимостей - чтобы мы вызвали у модуля `Init` и были уверены в том, что все зависимости будут прогружены и модуль будет готов к работе.
3. Результат инициализации - чтобы знать, что модуль инициализирован, а если нет, то в чем заключается ошибка.
4. Отслеживание состояния - чтобы знать текущее состояние модуля: еще не инициализирован, инициализирован, деинициализирован и т.д.

Это был небольшой `Specification`, надеюсь мои мысли и идеи совпадают с первоначальными.
Вот как примерно выглядит код инициализации и деинициализации модулей сейчас: 

https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_core.c#L765-L783

https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_core.c#L838-L853

Первоначально хотел показать реализацию непосредственно внедрив ее сразу, но столкнулся с тем, что не все модули имеют простой `Init` без аргументов - некоторые ожидают параметры, которые создаются на этапе других модулей, которые зависят от других и так далее...

Такая проблема есть в `vk_swapchain`, где инициализация ожидает внешние параметры:
https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_swapchain.c#L184

Эти параметры потом сохраняются в глобальное состояние:
https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_swapchain.c#L196-L197

Если посмотреть саму структуру где они хранятся, то видим комментарий, что они тут быть и не должны:
https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_swapchain.c#L9-L12

Сам модуль `vk_swapchain` инициализируется в модуле `vk_framectl` (кстати не знаю что за `framectl`, понятно лишь то, что `frame` - значит связано с кадрами, #709):
https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_framectl.c#L420-L432

Идем в заголовок и видим, что и здесь это не должно быть:
https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_framectl.h#L13-L22

Если посмотреть чуть выше, то видим комментарий, что тут вообще всех этих глобальных переменных не должно быть: 
https://github.com/w23/xash3d-fwgs/blob/a0b36a4301fbd718a4c3ae85bbe06d9efa50d6df/ref/vk/vk_framectl.h#L8C1-L8C1
```c
// TODO most of the things below should not be global. Instead, they should be passed as an argument/context to all the drawing functions that want this info
```
Такой вложенности мне хватило, чтобы понять, что сам я тут точно не разберусь...
Наверно это надо в отдельную ишью вынести, ну ладно, написал значит написал, может потом создам, где сразу для всех модулей выпишу эти зависимости.

Жду отзывов на данное творение и возможные предложения/исправления.

nilsonragee committed:
vk: create module API outline (WIP)

nilsonragee committed:
vk: create module API outline (WIP)

nilsonragee committed:
vk: modules: integrate `textures` module as an example

Also make a way to signal the module whether the shutdown was forced
or manual.

nilsonragee committed:
vk: modules: integrate `textures` module as an example

Also make a way to signal the module whether the shutdown was forced
or manual.

nilsoncore commented on ref/vk/vk_textures.c:
```diff
@@ -19,6 +19,15 @@
 #include <math.h> // sqrt
 
+RVkModule g_module_textures = {
+	.name = "textures",
```
Тут хорошо бы решить через что мы будем брать имя:
1. Через макрос `MODULE_NAME`.
2. Через поле `name`.

Мне кажется предпочтительнее 2-ое, потому что этот макрос можно использовать только внутри самого модуля. Если мы, например, захотим в одном модуле узнать имя другого модуля обратившись к нему, то через макрос это не выйдет.

nilsoncore commented on ref/vk/vk_module.h:
```diff
@@ -0,0 +1,97 @@
+
+#include "vk_core.h"
+
+typedef enum {
+	RVkModuleResult_Success = 0
+} RVkModuleResult;
```
Здесь идея в том, что мы не можем централизировать все ошибки в одном `enum`-е, т.к. мы попросту не можем знать все потенциальные `error` коды и определять их все тут - было бы глупо. Поэтому всё это делегируется модулям - они сами определяют возможные результаты.
Если что-то пошло не так, то чтобы посмотреть результат нужно будет привести тип к типу результатов этого модуля.
Например:
```c
RVkModuleResult textures_init = g_module_textures.Init( RVkModuleLogLevels_All, RVkModuleArgs_Empty );
if ( textures_init != RVkModuleResult_Success ) {
    RVkModuleTexturesResult textures_error = ( RVkModuleTexturesResult ) textures_init;
    if ( textures_error == RVkModuleTexturesResult_DefaultSamplerCreationError )
        // ...
    }
}
```
Да уж, устанешь писать, может быть стоит пересмотреть этот подход...
Но в целом идея в том, что мы не разбираемся что делать с ошибкой внутри модуля, а лишь возвращаем результат инициализации.
Возможно и нет смысла `if`-ами перебирать каждую ошибку и что-то делать, а просто вызвать `module.GetResultName( result )` и запринтить это, т.к. я предполагаю мы просто всё выключим и выйдем в любом случае.

nilsonragee committed:
Merge branch 'vk_modules' of https://github.com/nilsoncore/xash3d-fwgs into vk_modules

nilsonragee committed:
vk: modules: distinguish all possible modules and integrate them

nilsonragee committed:
vk: modules: fix minor bugs

Pass name of the module instead of the module itself in print args.
Forward declare `cl_entity_s` inside `camera.h`.

nilsonragee committed:
vk: modules: fix print format args

nilsonragee committed:
vk: modules: track init caller correctly

nilsoncore commented on ref/vk/vk_module.h:
```diff
@@ -42,56 +37,85 @@ enum {
-typedef RVkModuleResult ( *RVkModulePfn_Init )( RVkModuleLogLevels /*log_levels*/, RVkModuleArgs /*args*/ );
-typedef    const char * ( *RVkModulePfn_GetResultName )( RVkModuleResult /*result*/ );
-typedef            void ( *RVkModulePfn_Shutdown )( qboolean /*forced*/ );
+#define RVkModuleDependencies_Empty { .modules = NULL, .count = 0 }
+#define RVkModuleDependencies_FromStaticArray( array ) { .modules = array, .count = ( uint32_t ) COUNTOF( array ) }
```
Небольшие помогалки для указания зависимостей, чтобы вручную не писать каждый раз + можно было одним махом сделать глобальное изменение.

w23 replied:
(когда-нибудь потом) Здесь сразу напрашивается, конечно, некий ArrayView объект, т.к. это паттерн, который мы везде-везде тут используем.

nilsoncore commented on ref/vk/vk_module.h:
```diff
@@ -42,56 +37,85 @@ enum {
-	RVkModulePfn_Init          Init;
-	RVkModulePfn_GetResultName GetResultName;
-	RVkModulePfn_Shutdown      Shutdown;
+	RVkModule *init_caller;
```
Дополнительное поле чтобы отслеживать кто нас инициализировал. Используется в выводе (принтах).

nilsoncore commented on ref/vk/vk_overlay.c:
```diff
@@ -13,6 +13,32 @@
+static qboolean Impl_Init( void );
+static     void Impl_Shutdown( void );
+
+static RVkModule *required_modules[] = {
+	// Impl_Init: VK_BufferCreate
+	// Impl_Shutdown: VK_BufferDestroy
+	&g_module_buffer,
+
+	// drawFill: R_TextureFindByName
+	// R_DrawTileClear: R_TextureGetNameByIndex
+	&g_module_textures_api,
+
+	// drawOverlay: R_VkTextureGetDescriptorUnorm
+	// &g_module_textures, -- @TexturesInitHardcode
+
+	// createPipelines: VK_PipelineGraphicsCreate
+	&g_module_pipeline
+};
```
Не включаем в список `textures` (который `vk_textures`), чтобы его гарантированно инициализировал `texture_api` (который `r_textures`). Захардкоженый костыль.

nilsoncore commented on ref/vk/vk_overlay.c:
```diff
@@ -303,3 +306,38 @@ void CL_FillRGBABlend( float x, float y, float w, float h, int r, int g, int b,
+		return false;
+	}
+
+	// TODO this doesn't need to be host visible, could use staging too
+	if (!VK_BufferCreate("2d pics_buffer", &g2d.pics_buffer, sizeof(vertex_2d_t) * MAX_VERTICES,
+				VK_BUFFER_USAGE_VERTEX_BUFFER_BIT, VK_MEMORY_PROPERTY_HOST_COHERENT_BIT | VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT )) {
+		// FIXME cleanup -- NOTE(nilsoncore): Is it done?
+		gEngine.Con_Printf(S_ERROR "Couldn't create 2D pictures buffer.\n");
+		g_module_overlay.state = RVkModuleState_NotInitialized;
+		return false;
+	}
```
Опять-таки, непонятно как обрабатывать зачистку, если `VK_BufferCreate` всегда возвращает `true`, а в случае ошибки просто падает.

nilsoncore commented on ref/vk/vk_pipeline.c:
```diff
@@ -1,30 +1,39 @@
-{
-	vkDestroyPipelineCache(vk_core.device, g_pipeline_cache, NULL);
-}
+	// VK_PipelineGraphicsCreate: `.renderPass = vk_frame.render_pass.raster` -- @RenderpassOwnershipMess
+	// NOTE(nilsoncore): Because of this we can't include `framectl` as a dependency
+	// as it will create circular dependency chain: `render` -> `pipeline` -> `framectl` -> `render`.
+	// We have to assume that `framectl` is initialized by someone. Ugh.
+	// &g_module_framectl
```
Циркулярка, которую избежал костылем.

w23 replied:
Любопытно, как эта циркулярка вскрылась тут. Надо почесать репу 🤕. Не знаю пока, как решать.

nilsoncore commented on ref/vk/vk_ray_accel.c:
```diff
@@ -647,3 +627,73 @@ qboolean RT_BlasBuild(struct rt_blas_s *blas, const struct vk_render_geometry_s
+		R_SPEEDS_COUNTER(g_accel.stats.accels_built, "built", kSpeedsMetricCount);
+
+		gEngine.Con_Printf( S_NOTE "Module '%s' is initialized with enabled functionality.\n", g_module_ray_accel.name );
+	} else {
+		gEngine.Con_Printf( S_WARN "Module '%s' is initialized with disabled functionality.\n", g_module_ray_accel.name );
+		gEngine.Con_Printf( S_WARN "Switch to Ray Tracing renderer to enable it. (It must be supported on your graphics device)\n" );
+	}
```
См. комментарий в `rtx`.

w23 replied:
Как минимум индентация не нужна -- можно else clause сделать в начале и сделать там внутри return, тогда полезное тело функции останется с одинарной индентацией.

По-существу проблемы посмотрю только позже уже (два ночи тут), но пока кажется, что такого не должно быть.

nilsoncore commented on ref/vk/vk_rtx.c:
```diff
@@ -700,91 +744,104 @@ static void reloadPipeline( void ) {
+			"Fix random seed value for RT monte carlo sampling. Used for reproducible regression testing");
+
+		gEngine.Con_Printf( S_NOTE "Module '%s' is initialized with enabled functionality.\n", g_module_rtx.name );
+	} else {
+		gEngine.Con_Printf( S_WARN "Module '%s' is initialized with disabled functionality.\n", g_module_rtx.name );
+		gEngine.Con_Printf( S_WARN "Switch to Ray Tracing renderer to enable it. (It must be supported on your graphics device)\n" );
+	}
```
Я тут написал, что для полноценного включения функционала `rtx` нужно переключиться на RT рендер, НО - естественно никакой функционал переинициализации модуля я не писал, да и возможности проверить нет - у карточки нет `rayQuery`, а значит RT недоступен.

nilsoncore commented on ref/vk/vk_ray_model.c:
```diff
@@ -479,3 +473,52 @@ void RT_FrameAddOnce( rt_frame_add_once_t args ) {
+		}
+
+		gEngine.Con_Printf( S_NOTE "Module '%s' is initialized with enabled functionality.\n", g_module_ray_model.name );
+	} else {
+		gEngine.Con_Printf( S_WARN "Module '%s' is initialized with disabled functionality.\n", g_module_ray_model.name );
+		gEngine.Con_Printf( S_WARN "Switch to Ray Tracing renderer to enable it. (It must be supported on your graphics device)\n" );
+	}
```
См. комментарий в `rtx`.

nilsoncore commented on ref/vk/vk_render.c:
```diff
@@ -36,6 +36,48 @@
+	// VK_RenderEndRTX: VK_RayFrameEnd
+	&g_module_rtx,
+
+	// R_RenderModelCreate: RT_ModelCreate
+	// R_RenderModelDestroy: RT_ModelDestroy
+	// R_RenderModelUpdate: RT_ModelUpdate
+	// R_RenderModelUpdateMaterials: RT_ModelUpdateMaterials
+	// R_RenderModelDraw: RT_FrameAddModel
+	// R_RenderDrawOnce: RT_FrameAddOnce
+	// &g_module_ray_model, -- NOTE(nilsoncore): For some reason we don't see this module although we use its functions?? Whatever, it's inside rtx anyway.
```
Теперь я уже понял, что функции эти объявлены напрямую в `rtx`. Но `vk_ray_internal` я превратил в модуль `ray_model`...

w23 replied:
Да, в лучах тут некоторый беспорядок, который рано или поздно напрашивается на распутывание.

nilsoncore commented on ref/vk/vk_scene.c:
```diff
@@ -466,123 +517,6 @@ static int R_TransEntityCompare( const void *a, const void *b)
 	return 0;
 }
 
-// FIXME where should this function be
-#define RP_NORMALPASS() true // FIXME ???
-int CL_FxBlend( cl_entity_t *e ) // FIXME do R_SetupFrustum: , vec3_t vforward )
```
CL_FxBlend перенес в `camera`, по-моему ей там самое место. (нахождение ее в `scene` вызывало циркулярные зависимости)

w23 replied:
Бтв, это stateless функция, для неё не надо ничего инициализировать. Так что можно и на месте оставить.

nilsoncore commented on ref/vk/vk_sprite.c:
```diff
@@ -23,6 +25,40 @@
+	// R_DrawSpriteQuad: R_RenderModelDraw
+	&g_module_render,
+
+	// R_DrawSpriteQuad: R_VkMaterialModeFromRenderType -- @MaterialMode: Why is it in ray_model and not in render?
+	&g_module_ray_model,
```
Все-таки `ray_model` это больше как внутренний модуль `rtx`, а тут вот так, напрямую. Не очень как-то.

w23 replied:
`R_VkMaterialModeFromRenderType()` это stateless вспомогательная функция, для её использования не нужен модуль.

nilsoncore commented on ref/vk/vk_sprite.c:
```diff
@@ -1109,3 +1134,21 @@ void Mod_SpriteUnloadTextures( void *data )
+
+	XRVkModule_OnInitEnd( g_module_sprite );
+	return true;
+	// TODO return createQuadModel();
```
Это здесь было в оригинале. Не знаю что это значит, пока что оставил до выяснения обстоятельств.

w23 replied:
Там какая-то шляпа была с порядком инициализации, пусть пока будет, как есть.

nilsoncore commented on ref/vk/vk_textures.c:
```diff
@@ -58,104 +75,11 @@ static struct {
-
 static VkSampler pickSamplerForFlags( texFlags_t flags );
-static qboolean uploadTexture( int index, vk_texture_t *tex, rgbdata_t *const *const layers, int num_layers, qboolean cubemap, colorspace_hint_e colorspace_hint );
+static qboolean uploadTexture( int index, vk_texture_t *tex, const rgbdata_t *pic, colorspace_hint_e colorspace_hint );
```
Здесь просто обновление после `rebase`.

nilsoncore commented on ref/vk/vk_textures.c:
```diff
@@ -778,3 +648,85 @@ VkDescriptorImageInfo R_VkTexturesGetBlueNoiseImageInfo( void ) {
+static qboolean Impl_Init( void ) {
+	XRVkModule_OnInitStart( g_module_textures );
+
+	// NOTE(nilsoncore): We are expecting to be initialized only by main module. See @TexturesInitHardcode
+	if ( g_module_textures.init_caller != &g_module_textures_api ) {
+		ERR( "'%s' submodule can only be initialized by '%s'.", g_module_textures.name, g_module_textures_api.name );
+		g_module_textures.state = RVkModuleState_NotInitialized;
+		return false;
+	}
```
Ладно, получается `init_caller` не только для принтинга используется :Kappa:

nilsoncore commented on ref/vk/vk_triapi.c:
```diff
@@ -9,6 +9,21 @@
+static qboolean Impl_Init( void );
+static     void Impl_Shutdown( void );
+
+static RVkModule *required_modules[] = {
+	&g_module_sprite
+};
```
Похоже забыл указать где мы зовем функции из `sprite`.

nilsoncore commented on ref/vk/vk_triapi.c:
```diff
@@ -18,8 +33,6 @@ static struct {
 
 	vk_render_type_e render_type;
-
-	qboolean initialized;
```
`initialized` больше не нужен, т.к. состояние инициализации уже хранится внутри модуля.

nilsoncore commented on ref/vk/camera.c:
```diff
@@ -158,3 +158,120 @@ int TriWorldToScreen( const float *world, float *screen )
 	return retval;
 }
 
+// NOTE(nilsoncore): Moved from `vk_beams.c`.
+// It uses global `g_camera` and comment suggests to use `R_SetupFrustum` -- a function inside `camera.c` -- so probably it is a good place.
+#define RP_NORMALPASS() true // FIXME ???
+int CL_FxBlend( cl_entity_t *e ) // FIXME do R_SetupFrustum: , vec3_t vforward )
```
Перемещено из `scene` для избежания циклической зависимости, подробнее в `scene`.

nilsoncore commented on ref/vk/r_textures.c:
```diff
@@ -986,3 +923,105 @@ void R_TextureRelease( unsigned int texnum ) {
 	releaseTexture( texnum, ref_interface );
 }
+
+static qboolean Impl_Init( void ) {
```
Заранее извиняюсь за множество затронутых +- линий, которые на самом деле практически те же, просто все предыдущие функции `Init` и `Shutdown` перенесены в конец файла, т.к. до этого они постоянно были в разных местах.

nilsoncore commented on ref/vk/r_textures.c:
```diff
@@ -986,3 +923,105 @@ void R_TextureRelease( unsigned int texnum ) {
+
+	createDefaultTextures();
+
+	// NOTE(nilsoncore): This has to be hardcoded in a current approach. -- @TexturesInitHardcode
+	// They use each other's functions, so it will be a circular dependency if we include them in a normal way.
+	// We could instead not turn `textures` (vk_textures{h,c}) into a whole module and leave it as it is, 
+	// but it has `Init` and `Shutdown` functions like in a normal module, so it's bad either way.
+	// This probably needs some reorganization / change of logic.
+	g_module_textures.init_caller = &g_module_textures_api;
+	if ( !g_module_textures.Init() ) {
+		ERR( "Couldn't initialize '%s' submodule.", g_module_textures.name );
+		g_module_textures_api.state = RVkModuleState_NotInitialized;
+		return false;
+	}
```
`vk_textures` - это вроде бы просто внутренности `r_textures`, но у них есть свои `Init` и `Shutdown`, странно просто оставлять не тронутым. И, кстати, кто-то использует именно `vk_textures` без `r_speeds`, что странно. Не знаю, не хочется вот так оставлять файлы "в темноте", когда у них есть явный интерфейс инитов/деинитов, у них есть зависимости от других модулей и т.д. С другой стороны, будут возникать вот такие костыли. Наверно по итогу стоит реорганизовать компоненты так, чтобы такого не происходило в принципе.

nilsoncore commented on ref/vk/vk_core.c:
```diff
@@ -843,43 +804,34 @@ void R_VkShutdown( void ) {
-	R_VkCombuf_Destroy();
-
-	VK_DevMemDestroy();
+	g_module_brush.Shutdown();
+	g_module_overlay.Shutdown();
+	g_module_textures_api.Shutdown();
+	g_module_textures.Shutdown();
+	g_module_scene.Shutdown();
+	g_module_studio.Shutdown();
+	g_module_render.Shutdown();
+	g_module_geometry.Shutdown();
+	g_module_framectl.Shutdown();
+	g_module_pipeline.Shutdown();
+	g_module_staging.Shutdown();
+	g_module_buffer.Shutdown();
+
+	g_module_nv_aftermath.Shutdown();
+	g_module_descriptor.Shutdown();
+	g_module_devmem.Shutdown();
+	g_module_combuf.Shutdown();
```
Вот этого безобразия, конечно же, не должно быть - тут все то же самое, что и в предыдущем комментарии.
Однако тут дополнительный вопрос - должен ли модуль при своем `Shutdown` вызывать `Shutdown` зависимых модулей? Или модули сами должны выключаться по достижению `reference_count == 0`? Трекинг референсов я еще, кстати, не сделал.

Я пока не уверен как мы должны считать референсы (которые, насколько я понимаю, хранят количество внешних активных модулей, которые зависят от данного модуля?).

w23 replied:
В простом текущем варианте модуль должен делать "Acquire" своим зависимостями (refcount++ и инициализацию, если refcount был 0) при инициализации, и "Release", когда его гасят.
"Release" внутри себя уменьшает refcount, и делает настоящий Shutdown только при достижении им нуля.

Тут напрашивается где-то в самом конце `R_VkShutdown()` проверка на то, что все-все модули были погашены. Тут можно либо завести глобальный счётчик количества инициализированных модулей, либо глобальный список модулей. Счётчик проще сделать, и он будет железобетонный, но тогда надо будет грепать логи на предмет поиска непогашенного. В список можно будет забыть добавить и не обнаружить утечку.

nilsoncore commented:
Выкатил первую рабочую версию, НО - проверял только в растере, в RT рендере скорее всего не работает. Возможно оно будет работать, если происходит запуск сразу на RT, но переключение между рендерами точно работать не должно из-за логики работы связанных с RT модулями - подробнее в комментариях ревью.

Пока что все сообщения инитов / деинитов пишутся в консоль, но скорее всего нужно будет добавить что-то, что будет это контролировать, т.к. спамит достаточно. Хотя сразу видно какой модуль когда включен, когда выключен. Как это реализовать пока что не знаю - сделать ручную проверку аргумента типа `-vkdbgmod` или как-то через `vk_logs` реализовать, я пока что не разобрался и у меня уже просто мозг кипит от этого, потому что работал над этим целую неделю практически без перерыва. 

Если честно мне уже тошно от этого и пишу все это через силу, но я понимаю, что никто за меня ход моих мыслей и идей объяснять не будет, поэтому приходится. По сути большая часть работы сделана, осталось довести до конца, отполировать к финальному, готовому виду.

К сожалению коммиты читать очень тяжело, потому что они в спаме изменений из-за того, что я все функции инитов и деинитов перенес в конец файлов, т.к. они в разных файлах были в разных местах, плюс если они в конце файла, то никакие лишние форвард-декларации плодить не придется. Во всяком случае постарался писать комментарии в ревью под значимыми кусками кода, чтобы не потеряться в спаме.

На этом, наверно, пока все - буду ждать разбора на стриме. А до этого, я наверно немного развеюсь, потому что реально уже голова кипит мало того, что от комплексности самой проблемы и методов реализаций, так еще и от того, что это нужно все объяснять... В общем, надеюсь меня поняли.

nilsoncore commented:
Мержить пока нельзя, есть еще косяки - просто жду ревью текущих изменений.

nilsoncore commented:
Т.к. во вкладке измененных файлов будет много лишних коммитов, в том числе не моих из-за `rebase` до текущего состояний ветки, то вот небольшой гайд как выбрать конкретные коммиты:

![image](https://github.com/w23/xash3d-fwgs/assets/65307695/1adfc671-a4dd-47c0-a892-ac760826dcbf)

Для этого нужно зайти во вкладку коммитов и пролистать в самый низ (новые коммиты начинаются снизу).
Дальше советую выбрать коммиты от самого нового до `vk: modules: distinguish all possible modules` (`99d3fc5`):
для этого нужно с зажатым шифтом сначала нажать на самый нижний коммит, а затем на тот, до которого будем смотреть изменения - во время выделения они будут подсвечиваться желтым.

![image](https://github.com/w23/xash3d-fwgs/assets/65307695/d231fc75-6f19-4425-a0d7-6131a5aa9a4e)

w23 commented:
Это гигантская работа, спасибо!
У меня займёт сколько-то часов её переварить. Буду оставлять комментарии по мере того, как появляется время переваривать.

w23 commented on ref/vk/vk_brush.c:
```diff
@@ -1909,3 +1947,25 @@ void R_BrushUnloadTextures( model_t *mod )
 }
+
+static qboolean Impl_Init( void ) {
+	XRVkModule_OnInitStart( g_module_brush );
```
Пока не понял, зачем нужны эти особые виньетки каждый раз. Если уж Init/Shutdown зовутся снаружи, то пускай они там один раз в подобное и будут обёрнуты.

w23 commented on ref/vk/vk_buffer.c:
```diff
@@ -17,6 +34,8 @@ qboolean VK_BufferCreate(const char *debug_name, vk_buffer_t *buf, uint32_t size
 		memreq.alignment = ALIGN_UP(memreq.alignment, vk_core.physical_device.properties_ray_tracing_pipeline.shaderGroupBaseAlignment);
 	}
+
+	ASSERT( g_module_devmem.state == RVkModuleState_Initialized && "VK_BufferCreate: 'devmem' is not initialized." );
```
По идее такое должно делаться внутри самого devmem. Проверка руками снаружи лишняя -- всё равно будет хлопок, если оно не инициализировано правильно.

nilsoncore replied:
Это я тут выяснял причину крашей, забыл убрать.

w23 commented on ref/vk/vk_core.c:
```diff
@@ -770,70 +734,67 @@ qboolean R_VkInit( void )
 		return false;
 	}
 
-#if USE_AFTERMATH
-	if (!VK_AftermathInit()) {
-		gEngine.Con_Printf( S_ERROR "Cannot initialize Nvidia Nsight Aftermath SDK\n" );
-	}
-#endif
-
```
Думаю, что aftermath это пока "особый случай", наравне с инициализацией вулкан инстанса и устройства. Он может оставаться ровно таким, как был до модулей, можно вообще не трогать пока.

w23 commented on ref/vk/vk_core.c:
```diff
@@ -770,70 +734,67 @@ qboolean R_VkInit( void )
-		VK_LightsInit();
+	/* Print all modules with their dependencies (not printing states) */
+
+	RVkModule_PrintDependencyList( &g_module_combuf, false );
+	RVkModule_PrintDependencyList( &g_module_buffer, false );
+	RVkModule_PrintDependencyList( &g_module_devmem, false );
+	RVkModule_PrintDependencyList( &g_module_nv_aftermath, false );
+	RVkModule_PrintDependencyList( &g_module_staging, false );
+	RVkModule_PrintDependencyList( &g_module_image, false );
+	RVkModule_PrintDependencyList( &g_module_pipeline, false );
+	RVkModule_PrintDependencyList( &g_module_descriptor, false );
+	RVkModule_PrintDependencyList( &g_module_framectl, false );
+	RVkModule_PrintDependencyList( &g_module_geometry, false );
+	RVkModule_PrintDependencyList( &g_module_render, false );
+	RVkModule_PrintDependencyList( &g_module_studio, false );
+	RVkModule_PrintDependencyList( &g_module_scene, false );
+	RVkModule_PrintDependencyList( &g_module_textures_api, false );
+	RVkModule_PrintDependencyList( &g_module_textures, false );
+	RVkModule_PrintDependencyList( &g_module_overlay, false );
+	RVkModule_PrintDependencyList( &g_module_brush, false );
+	RVkModule_PrintDependencyList( &g_module_rtx, false );
+	RVkModule_PrintDependencyList( &g_module_light, false );
+	RVkModule_PrintDependencyList( &g_module_sprite, false );
+	RVkModule_PrintDependencyList( &g_module_beams, false );
```
На время разработки пусть будет, но в конце перед мержем у нас тут по идее останется всего несколько модулей (в идеале два-три: рендер, лучи, оверлей; но пока хз, как сложится), вот для них можно будет и вывести все зависимости деревом.

w23 commented on ref/vk/vk_core.c:
```diff
@@ -770,70 +734,67 @@ qboolean R_VkInit( void )
+	RVkModule_PrintDependencyList( &g_module_sprite, false );
+	RVkModule_PrintDependencyList( &g_module_beams, false );
+
+	/* Modules without dependencies */
+
+	if ( !g_module_combuf.Init() )  return false;
+	if ( !g_module_devmem.Init() )  return false;
+	if ( !g_module_descriptor.Init() )  return false;
+	if ( !g_module_nv_aftermath.Init() )  return false;
+
+	/* Modules with dependencies */
+
+	if ( !g_module_buffer.Init() )  return false;
+	if ( !g_module_staging.Init() )  return false;
+	if ( !g_module_image.Init() )  return false;
+	if ( !g_module_pipeline.Init() )  return false;
+	if ( !g_module_framectl.Init() )  return false;
+	if ( !g_module_geometry.Init() )  return false;
+	if ( !g_module_render.Init() )  return false;
+	if ( !g_module_studio.Init() )  return false;
+	if ( !g_module_scene.Init() )  return false;
+	if ( !g_module_textures_api.Init() )  return false; // +g_module_textures (hardcoded) -- @TexturesInitHardcode
+	if ( !g_module_overlay.Init() )  return false;
+	if ( !g_module_brush.Init() )  return false;
```
Пока не нужно. На текущем этапе в модули надо сконвертить только большую часть того, для чего тут зовутся Init/Shutdown. Что конвертится с трудом, или плохо/нетривиально мапится на модули, можно не трогать.

Разбиение на более атомарные модули может прагматично понадобиться (или не понадобиться!) позже, с опытом работы с модулями. С той горы виднее будет, но пока мы на неё не взошли.

w23 commented on ref/vk/vk_framectl.c:
```diff
@@ -19,6 +19,47 @@
+	// R_BeginFrame: R_VkSwapchainAcquire
+	// submit: R_VkSwapchainPresent
+	// Impl_Init: R_VkSwapchainSetRenderPassAndDepthFormat_FIXME -- @RenderpassOwnershipMess
+	&g_module_swapchain,
```
swapchain не зовётся напрямую из vk_core, можно пока не выносить в отдельный модуль

w23 commented on ref/vk/vk_framectl.c:
```diff
@@ -762,3 +736,83 @@ qboolean VID_ScreenShot( const char *filename, int shot_type )
+	if (vk_core.rtx)
+		vk_frame.render_pass.after_ray_tracing = createRenderPass(depth_format, true);
+
+	// Render pass and depth format have to be set before we begin to init swapchain.
+	// (Swapchain will be initialized as a dependency/required module)
+	R_VkSwapchainSetRenderPassAndDepthFormat_FIXME( vk_frame.render_pass.raster, depth_format ); // -- @RenderpassOwnershipMess
```
Сложно слишком. Можно пока оставить, как было.

w23 commented on ref/vk/vk_geometry.c:
```diff
@@ -175,13 +199,27 @@ void R_GeometryBuffer_MapClear( void ) {
 }
 
-qboolean R_GeometryBuffer_Init(void) {
+void R_GeometryBuffer_Flip(void) {
+	R_BlocksClearOnce(&g_geom.alloc);
+}
+
+VkBuffer R_GeometryBuffer_Get(void) {
+	return g_geom.buffer.buffer;
+}
+
+qboolean Impl_Init( void ) {
+	XRVkModule_OnInitStart( g_module_geometry );
+
```
Вот пошли песни и пляски с диффом от переноса.

w23 commented on ref/vk/vk_image.c:
```diff
@@ -8,6 +8,28 @@
+	.name = "image",
+	.state = RVkModuleState_NotInitialized,
+	.dependencies = RVkModuleDependencies_FromStaticArray( required_modules ),
+	.Init = Impl_Init,
+	.Shutdown = Impl_Shutdown
```
Бтв можно просто `NULL` для функций-пустышек делать.

w23 commented on ref/vk/vk_light.c:
```diff
@@ -38,6 +38,30 @@
+	&g_module_staging
+};
+
+RVkModule g_module_light = {
+	.name = "light",
```
Да тут такого в кодобазе полно, она очень органическая без ГМО

w23 commented on ref/vk/vk_light.c:
```diff
@@ -1394,3 +1384,46 @@ void RT_LightsFrameEnd( void ) {
+
+	const int buffer_size = sizeof(struct LightsMetadata) + sizeof(struct LightCluster) * MAX_LIGHT_CLUSTERS;
+
+	if (!VK_BufferCreate("rt lights buffer", &g_lights_.buffer, buffer_size,
+		VK_BUFFER_USAGE_STORAGE_BUFFER_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT,
+		VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT)) {
+		// FIXME complain, handle -- NOTE(nilsoncore): Is it done?
+		ERR( "Couldn't create Ray Tracing lights buffer." );
+		g_module_light.state = RVkModuleState_NotInitialized;
+		return false;
+	}
```
> Здесь было просто `return false`, добавил немного обработки и спросил в комментариях - та ли эта обработка, которая здесь ожидалась быть? 

Я бы ограничил скоуп этого PR только модулями. Обработка ошибок и прочее это отдельный тяжёлый PR, лучше их не мешать. Там такого во многих местах очень много.

> Кстати немного странная проверка, потому что `VK_BufferCreate` не может вернуть `false`, вместо этого он просто свалится, вроде.

И это тоже. Надо смотреть отдельно. Пока мы просто мрём на кучу ошибок, тогда как надо уметь gracefully гаситься и фоллбечиться на GL рендер. Это всё сильно-сильно потом, даже совсем после релиза.

w23 commented on ref/vk/vk_module.c:
```diff
@@ -0,0 +1,150 @@
+// This approach is quite simple and clean: you just write this construction passing it module and that's it.
+// You don't have to do anything else and know how it works inside (for the most part).
+// Although, it has a downside: these macros are CHUNKY. We don't just call some function, we paste the whole code inline
+// for every module out there.  It probably grows the binary size a lot (didn't measure).
```
oh no, my internal sizecoder is in tears

w23 commented on ref/vk/vk_module.c:
```diff
@@ -1,92 +1,150 @@
-		module_index += 1;
+#include "debugbreak.h"
+
+// NOTE(nilsoncore):
+// These `replica` functions are broken for now because I didn't update them to be exact as `XRVkModule_Onxxx` macros.
+// The reason why these functions exist (for now) is that I don't know whether we should use macros or functions like these.
+// For now we just wrap around every `Impl_Init` module function with `XRVkModule_OnxxxStart` and `XRVkModule_OnxxxEnd`.
+// 
+// This approach is quite simple and clean: you just write this construction passing it module and that's it.
+// You don't have to do anything else and know how it works inside (for the most part).
+// Although, it has a downside: these macros are CHUNKY. We don't just call some function, we paste the whole code inline
+// for every module out there.  It probably grows the binary size a lot (didn't measure).
+// 
+// The second approach is to use regular functions.  It is probably more conventional and rational way, but it has downsides too.
+// You can't just write it and expect it to work.  The placing is the same as with the macros, but things start to become tricky.
+// We can't use `qboolean` as a result value because of the `already initialized?` check.
+// If module is already initialized we return `true` from the module `Impl_Init` function.  Here inside `OnInitStart` it's too ambigious -
+// does that mean that module is already initialized and we are done or module has loaded dependencies and ready to initialize?
+// In this case we should probably declare some enum with states like `_AlreadyInitialized`, `_DependenciesLoaded`, `_FailedToLoadDependecies`, etc.
+// In the end we will end up not just with the boilerplate code that we have to paste everywhere, but also we can't make change in one place --
+// every change must be then made manually in every file.
+// 
+// Whether we want it or not -- we must decide which approach we will use. Don't be afraid to comment on this if I missed something.
```
Я пока вообще не понял, зачем они нужны в любом виде, где мы явно их должны руками дёргать в начале и конце.
То, что эти функции должны делать, решается абсолютно вне Init/Shutdown ровно одной копией кода внутри vk_module.c, который резолвит зависимости и прочее. Init/Shutdown уже зовутся только оттуда.
Это неявно подразумевает то, что никто и никогда не зовёт Init/Shutdown модулям напрямую -- это делает только vk_module.c под капотом, т.е. создание/гашение модулей это, напрмер, `R_ModuleInit(&g_module_framectl)`...

w23 commented on ref/vk/vk_module.c:
```diff
@@ -0,0 +1,150 @@
+	qboolean second_try = false;
+	uint32_t deps_in_progress_count = 0;
+
+init_dependencies:
```
Неуместное использование label+goto, здесь лучше использовать именно цикл, раз внизу к тому же используется счётчик.
Единственное известное уместное употребление goto в этой кодобазе это чистка за собой в конце функции.

Также, пока не понял, зачем нам более одной попытки инициализации вообще. Пока представляется, что это сразу фатальная ошибка, если не удалось с первого раза поднять.

w23 commented on ref/vk/vk_pipeline.c:
```diff
@@ -1,30 +1,39 @@
-		.sType = VK_STRUCTURE_TYPE_PIPELINE_CACHE_CREATE_INFO,
-		.initialDataSize = 0,
-		.pInitialData = NULL,
-	};
+static RVkModule *required_modules[] = {
+	// VK_PipelineRayTracingTrace: R_VkCombufScopeBegin, R_VkCombufScopeEnd
+	&g_module_combuf,
```
во дифу башню снесло от переноса функций. не надо переносить функции

w23 commented on ref/vk/vk_ray_internal.h:
```diff
@@ -11,6 +13,8 @@
 
 #include "shaders/ray_interop.h"
 
+extern RVkModule g_module_ray_model;
```
Тут по идее нужно только динамический `g_module_rt_dynamic_model` сделать модулем.

w23 reviewed (changes requested):
Так, я очень устал и сделал только %40 ревью.
Это очень большая работа, очень спасибо!

Пока что самые главные замечания это:
1. Не надо впредь делать rebase. Этот очень грубый инструмент сносит башню всему на свете. Rebase можно делать только крайне аккуратно на личных work-in-progress ветках, которые никому не показываются, никуда не пушатся, и тем более не участвуют в PR. Нам скорее всего придётся в итоге закрыть этот PR, создать новую ветку через ручное протаскивание изменений, и открыть новый PR.
2. Не надо переносить функции Init/Shutdown и трогать их внутренности. Это раздувает PR и делает его нечитаемым. Достаточно просто добавить RVkModule структуру в самый конец файла, ссылаясь на существующие имена функций.
3. Не надо переименовывать функции. Не уникальные имена функций значительно усложнают грепанье кода.
4. Не надо разбивать на мелкие модули. Не каждый файл или сущность модуль. Модули это более-менее то, что имеет Init/Shutdown пару в vk_core, с единичными исключениями.

w23 commented:
Йоу @nilsoncore чё каво?
Я тут потихоньку восстанавливаюсь-выкарабкиваюсь, и думаю, что вот это всё тут через недельку-другую мне может понадобиться.
Если у тебя нет времени на это глянуть, могу ли я взять твои наработки, и допилить их сам?

nilsoncore commented:
Йоу @w23

Решил я таки добить этот PR, но в текущем виде скорее всего это уже невозможно - услышал где-то про элегантность ребейза и зафакапил тут все. Скорее всего придется создать новый и ссылаться на этот для истории обсуждения.

Если честно, придется заново пробежаться по всему, что я тут наплодил и вникнуть, благо комментарии вроде обширные оставлял. Комментарии все прочитал и учту в новом PR, постараюсь не раздувать лишним.

> Я тут потихоньку восстанавливаюсь-выкарабкиваюсь, и думаю, что вот это всё тут через недельку-другую мне может понадобиться.
> Если у тебя нет времени на это глянуть, могу ли я взять твои наработки, и допилить их сам?

Как сказал выше, думаю все же допилить, но если сильно надо и есть прям четкое видение как все это реализовать, то конечно, я не против, могу заняться чем-то другим, потому что пока насчет финального вида я на 100% не уверен, плюс забыл все, надо вспоминать.

w23 commented:
> Решил я таки добить этот PR, но в текущем виде скорее всего это уже невозможно - услышал где-то про элегантность ребейза и зафакапил тут все. Скорее всего придется создать новый и ссылаться на этот для истории обсуждения.
> 
> Если честно, придется заново пробежаться по всему, что я тут наплодил и вникнуть, благо комментарии вроде обширные оставлял. Комментарии все прочитал и учту в новом PR, постараюсь не раздувать лишним.

Ага, мы тут все выпали и потеряли контекст.


> > Я тут потихоньку восстанавливаюсь-выкарабкиваюсь, и думаю, что вот это всё тут через недельку-другую мне может понадобиться.
> > Если у тебя нет времени на это глянуть, могу ли я взять твои наработки, и допилить их сам?
> 
> Как сказал выше, думаю все же допилить, но если сильно надо и есть прям четкое видение как все это реализовать, то конечно, я не против, могу заняться чем-то другим, потому что пока насчет финального вида я на 100% не уверен, плюс забыл все, надо вспоминать.

Я всё ещё не смотрел на это внимательно, и абсолютно чёткого видения нет. Но емнип я тут в комментариях ревью накидывал примерный вектор движения, и там начинали вырисовываться очертания желаемого.

w23 commented:
Так, пролистал комментарии, и ничего там не вырисовывается. Вернее, оно вырисовывается, но только у меня в голове, никуда напоказ я это не вывалил.

Предлагаю попробовать пройтись тут в комментах через лёгкое дизайн-ревью, т.е. прежде чем писать какой-то код, примерно накидать требования и обсудить апи. И только потом восстанавливать подходящие куски кода из этой ветки в свежую.

## Например, сыро и тезисами, что хотим:
### Зависимости
- ЧТО: В более явном виде указывать зависимости между модулями. Модуль при инициализации дёргает за другие модули, указывая, что они ему нужны.
- ЗАЧЕМ: Автоматический и менее хрупкий порядок инициализации -- сейчас там аккуратно выстроена последовательность руками. Ловить кольцевые зависимости рано.
- КАК:  Можно из модульного Init дёргать зависимости напрямую, не усложняя какой-то специальной структурой.

### Времена жизни модулей
- ЧТО: Модуль должен деинициализироваться, когда не нужен.
- ЗАЧЕМ: Опять же, чтобы всё чистилось само и вовремя
- КАК: Refcount. Модуль берётся через Acquire, высвобождается через Release.

### Валидация (опционально)
- ЧТО: Ловить использование неинициализированных модулей 
- ЗАЧЕМ: Проверка, что ничего не ломаем.
- КАК: Пока неясно, как именно. Можно руками рассыпать проверку на инициализацию в модульных функциях, но это может давать false negative, если кто-то случайно уже проинициализировал.

#### Аппендикс
Тут, кстати, можно озадачиться тем, чтобы перевести модули со статической модели (где у них есть `static struct { .. } g;`) на динамическую, где мы этот `g` маллочим во время инициализации.

Почему:
- Придётся явно таскать за собой указатель/handle на модуль, что помогает с валидацией: нам теперь НУЖНО где-то брать этот указатель.
- Явная динамическая аллокация упрощает слежение за потреблением памяти (в будущем).

## Как делать
1. Придумать апи и согласиться <-- мы находимся здесь.
2. Итеративно -- отдельными PR переводить модули на новую схему по одному. Зачем так: чтобы не плодить megachonker. Можно делать это отдельными PR в ветку с модулями, которая потом большим PR заедет в vulkan. Зачем может быть так нужно: в итеративном процессе может оказаться, что мы что-то придумали не так, и надо поменять. И можно будет это сделать, не ломая vulkan.
